plugins/builds/update.js

```diff
@@ -185,20 +182,26 @@ async function getStage({ stageFactory, workflowGraph, jobName, pipelineId }) {
  * Checks if all builds in stage are done running
  * @param  {Object}     stage                     Stage
  * @param  {Object}     event                     Event
+ * @param  {String}     buildStatus               Current build status
  * @return {Boolean}              Flag if stage is done
  */
-async function isStageDone({ stage, event }) {
-    // Get all jobIds for jobs in the stage
-    const stageJobIds = stage.jobIds;
+async function isStageDone({ stage, event, buildStatus }) {
+    // Combine job IDs in the stage and the stage setup ID
+    const stageJobIds = [...stage.jobIds, stage.setup];
+    // Get all builds for jobs in the stage for this event
+    const stageJobBuilds = await event.getBuilds({ params: { jobId: stageJobIds } });
 
-    stageJobIds.push(stage.setup);
+    // If no builds exist, return false
+    if (!stageJobBuilds || stageJobBuilds.length === 0) {
+        return false;
+    }
 
-    // Get all builds in a stage for this event
-    const stageJobBuilds = await event.getBuilds({ params: { jobId: stageJobIds } });
-    let stageIsDone = false;
+    // Check if all builds in the stage are finished
+    const stageIsDone = !stageJobBuilds.some(b => !FINISHED_STATUSES.includes(b.status));
 
-    if (stageJobBuilds && stageJobBuilds.length !== 0) {
-        stageIsDone = !stageJobBuilds.some(b => !FINISHED_STATUSES.includes(b.status));
+    // If the build status is 'SUCCESS', ensure all jobs are completed
+    if (buildStatus === 'SUCCESS') {
+        return stageJobBuilds.length === stageJobIds.length && stageIsDone;
```

Comment:
I believe we should not be relying on the count of stage job builds.

For Ex: Let us consider a stage which has 5 jobs `jobA`, `jobB`, `jobC1`, `jobC2` and `jobD`. `jobA`, `jobB`, `jobC1`, `jobD` are triggered in parallel after stage setup is completed. `jobC2` is downstream of `jobC1`. Let us assume `jobA`, `jobB` and `jobC1` have completed with statuses shown below and the current build is for `jobD` with status `SUCCESS`.
| Job | Build Status |
| --- | --- |
| setup | SUCCESS |
| jobA | SUCCESS |
| jobB | SUCCESS |
| jobC1 | FAILURE |
| jobD | SUCCESS |
There will no active builds (There is be no build for `jobC2`). Per this logic, `stageJobBuilds.length` (5) would not match `stageJobIds.length` (6). Stage will never be marked as done and `teardown` may not get triggered.